### 1. Loading and Preprocessing

In [29]:
import pandas as pd
from sklearn.datasets import fetch_california_housing
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score


data = fetch_california_housing()
df = pd.DataFrame(data.data, columns=data.feature_names)
df["Target"] = data.target

df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [30]:
df.isnull().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
Target        0
dtype: int64

In [31]:
x = df.drop("Target", axis=1)
y = df["Target"]

In [32]:
# Feature scaling
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

print(x_scaled)

[[ 2.34476576  0.98214266  0.62855945 ... -0.04959654  1.05254828
  -1.32783522]
 [ 2.33223796 -0.60701891  0.32704136 ... -0.09251223  1.04318455
  -1.32284391]
 [ 1.7826994   1.85618152  1.15562047 ... -0.02584253  1.03850269
  -1.33282653]
 ...
 [-1.14259331 -0.92485123 -0.09031802 ... -0.0717345   1.77823747
  -0.8237132 ]
 [-1.05458292 -0.84539315 -0.04021111 ... -0.09122515  1.77823747
  -0.87362627]
 [-0.78012947 -1.00430931 -0.07044252 ... -0.04368215  1.75014627
  -0.83369581]]


### Preprocessing steps performed

#### a.Converted the dataset into a Pandas DataFrame 
#### b. checked for misssing values.
#### c. separated target and features.
#### d.  Feature scaling
     

### 2. Regression Algorithm Implementation

In [33]:
#linear regression
from sklearn.linear_model import LinearRegression
x_train, x_test, y_train, y_test = train_test_split(
    x_scaled, y, test_size=0.2, random_state=42
)

model = LinearRegression()

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

#### Linear Regression finds the best-fitting line that represents the relationship between the input variables and the target value.
#### California Housing has several numerical features such as income, house age, number of rooms, population, latitude etc... Linear Regression can use all these features to predict house values and provides a simple baseline model.

In [34]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor

model = DecisionTreeRegressor(random_state=42)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

#### A Decision Tree Regressor makes a series of if-else decisions to divide the data into smaller groups.The California Housing dataset contains relationships that may not be purely linear. A Decision Tree can capture non-linear relationships and interactions between different housing features.

In [35]:
#random Forest Regressor
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

#### Random Forest creates many Decision Trees instead of using just one tree.Each tree makes a prediction, and the Random Forest combines their predictions, usually by taking their average.

In [36]:
#Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor

model = GradientBoostingRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

#### Gradient Boosting builds trees one after another. Each new tree tries to correct the errors made by the previous trees.

In [37]:
#Support Vector Regression (SVR)
from sklearn.svm import SVR

model = SVR(kernel="rbf")

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

#### SVR tries to find a function that predicts the target values while keeping most predictions within a specified margin of error.

### 3. Model Evaluation and Comparison

In [39]:
#Linear Regression
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Linear Regression")
print("MSE:", mean_squared_error(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Linear Regression
MSE: 0.5558915986952442
MAE: 0.5332001304956565
R²: 0.575787706032451


In [40]:
#Decision Tree
from sklearn.tree import DecisionTreeRegressor

model = DecisionTreeRegressor(random_state=42)
model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Decision Tree Regressor")
print("MSE:", mean_squared_error(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Decision Tree Regressor
MSE: 0.4942716777366763
MAE: 0.4537843265503876
R²: 0.6228111330554302


In [41]:
#Random forest
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Random Forest Regressor")
print("MSE:", mean_squared_error(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Random Forest Regressor
MSE: 0.25549776668540763
MAE: 0.32761306601259704
R²: 0.805024407701793


In [42]:
#Gradient Boosting
from sklearn.ensemble import GradientBoostingRegressor

model = GradientBoostingRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Gradient Boosting Regressor")
print("MSE:", mean_squared_error(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Gradient Boosting Regressor
MSE: 0.29399901242474274
MAE: 0.37165044848436773
R²: 0.7756433164710084


In [44]:
#SVR
from sklearn.svm import SVR

model = SVR(kernel="rbf")

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Support Vector Regressor")
print("MSE:", mean_squared_error(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Support Vector Regressor
MSE: 0.3551984619989416
MAE: 0.3977630963437863
R²: 0.7289407597956463


#### Based on the evaluation results, the Random Forest Regressor performed the best among the five regression algorithms. It achieved the lowest MSE (0.2555) and MAE (0.3276), along with the highest R² score (0.8050).

#### The Linear Regression model was the worst-performing algorithm among the five models. It had the highest MSE (0.5559) and highest MAE (0.5332), along with the lowest R² score (0.5758).